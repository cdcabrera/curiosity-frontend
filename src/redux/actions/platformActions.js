import {
  addNotification as RcsAddNotification,
  removeNotification as RcsRemoveNotification,
  clearNotifications as RcsClearNotifications
} from '@redhat-cloud-services/frontend-components-notifications';
import { platformTypes } from '../types';
import { platformServices } from '../../services/platform/platformServices';
import { translate } from '../../components/i18n/i18n';

/**
 * Platform service wrappers for dispatch, state update.
 *
 * @memberof Actions
 * @module PlatformActions
 */

/**
 * Add a platform plugin toast notification.
 *
 * @param {object} data
 * @returns {*}
 */
const addNotification = data => RcsAddNotification(data);

/**
 * Remove a platform plugin toast notification.
 *
 * @param {string} id
 * @returns {*}
 */
const removeNotification = id => RcsRemoveNotification(id);

/**
 * Clear all platform plugin toast notifications.
 *
 * @returns {*}
 */
const clearNotifications = () => RcsClearNotifications();

/**
 * Get an emulated and combined API response from the platforms "getUser" and "getUserPermissions" global methods.
 *
 * @param {string|Array} appName
 * @returns {Function}
 */
const authorizeUser = appName => dispatch =>
  dispatch({
    type: platformTypes.PLATFORM_USER_AUTH,
    payload: Promise.all([platformServices.getUser(), platformServices.getUserPermissions(appName)])
  });

/**
 * Return a "dispatch ready" export poll status check.
 *
 * @param {Function} dispatch
 * @returns {Function}
 */
const setExportStatus =
  dispatch =>
  (id, success = {}, error) =>
    dispatch({
      type: platformTypes.SET_PLATFORM_EXPORT_STATUS,
      payload: (error && Promise.reject(error)) || Promise.resolve(success),
      meta: {
        id
      }
    });

/**
 * Create an export status poll with download, and toast notifications.
 *
 * @param {object} options Apply polling options
 * @returns {Function}
 */
const getExistingExports =
  (options = {}) =>
  dispatch => {
    dispatch({
      type: platformTypes.SET_PLATFORM_EXPORT_STATUS,
      payload: platformServices.getExistingExports(undefined, {
        ...options,
        poll: {
          ...options.poll,
          status: (successResponse, ...args) => {
            if (!successResponse?.data?.data?.isAnythingPending && successResponse?.data?.data?.isAnythingCompleted) {
              dispatch(removeNotification('swatch-global-export'));
              dispatch(
                addNotification({
                  id: 'swatch-global-export',
                  title: translate('curiosity-toolbar.notifications', {
                    context: ['export', 'completed', 'titleGlobal']
                  }),
                  description: translate('curiosity-toolbar.notifications', {
                    context: ['export', 'completed', 'descriptionGlobal']
                  }),
                  dismissable: true
                })
              );
            }

            setExportStatus(dispatch)('global', successResponse, ...args);
          }
        }
      }),
      meta: {
        id: 'global',
        notifications: {
          rejected: {
            variant: 'warning',
            title: translate('curiosity-toolbar.notifications', {
              context: ['export', 'error', 'title']
            }),
            description: translate('curiosity-toolbar.notifications', {
              context: ['export', 'error', 'description']
            }),
            dismissable: true
          },
          pending: {
            id: 'swatch-global-export',
            title: translate('curiosity-toolbar.notifications', {
              context: ['export', 'pending', 'titleGlobal']
            }),
            description: translate('curiosity-toolbar.notifications', {
              context: ['export', 'pending', 'descriptionGlobal']
            }),
            dismissable: true
          }
        }
      }
    });
  };

/**
 * Create an export for download with toast notifications.
 *
 * @param {string} id
 * @param {object} data
 * @param {object} options Apply polling options
 * @returns {Function}
 */
const createExport =
  (id, data = {}, options = {}) =>
  dispatch => {
    dispatch({
      type: platformTypes.SET_PLATFORM_EXPORT_CREATE,
      payload: platformServices.postExport(data, {
        ...options,
        poll: {
          ...options.poll,
          status: (successResponse, ...args) => {
            if (
              !successResponse?.data?.data?.products?.[id]?.isPending &&
              successResponse?.data?.data?.products?.[id]?.isCompleted
            ) {
              dispatch(removeNotification('swatch-create-export'));
              dispatch(
                addNotification({
                  variant: 'success',
                  id: 'swatch-create-export',
                  title: translate('curiosity-toolbar.notifications', {
                    context: ['export', 'completed', 'title']
                  }),
                  description: translate('curiosity-toolbar.notifications', {
                    context: ['export', 'completed', 'description'],
                    fileName: successResponse?.data?.data?.products?.[id]?.completed?.[0]?.fileName
                  }),
                  dismissable: true
                })
              );
            }

            setExportStatus(dispatch)(id, successResponse, ...args);
          }
        }
      }),
      meta: {
        notifications: {
          rejected: {
            variant: 'warning',
            title: translate('curiosity-toolbar.notifications', {
              context: ['export', 'error', 'title']
            }),
            description: translate('curiosity-toolbar.notifications', {
              context: ['export', 'error', 'description']
            }),
            dismissable: true
          },
          pending: {
            id: 'swatch-create-export',
            title: translate('curiosity-toolbar.notifications', {
              context: ['export', 'pending', 'title', id]
            }),
            dismissable: true
          }
        }
      }
    });
  };

/**
 * Hide platform global filter.
 *
 * @param {boolean} isHidden
 * @returns {{Function}}
 */
const hideGlobalFilter = isHidden => ({
  type: platformTypes.PLATFORM_GLOBAL_FILTER_HIDE,
  payload: platformServices.hideGlobalFilter(isHidden)
});

const platformActions = {
  addNotification,
  removeNotification,
  clearNotifications,
  authorizeUser,
  createExport,
  setExportStatus,
  getExistingExports,
  hideGlobalFilter
};

export {
  platformActions as default,
  platformActions,
  addNotification,
  removeNotification,
  clearNotifications,
  authorizeUser,
  createExport,
  setExportStatus,
  getExistingExports,
  hideGlobalFilter
};
