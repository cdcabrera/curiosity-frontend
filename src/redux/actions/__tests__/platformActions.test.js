import { platformActions } from '../platformActions';

describe('PlatformActions', () => {
  it('Should return a dispatch object for the initializeChrome method', () => {
    expect(platformActions.initializeChrome()).toMatchSnapshot('dispatch object');
  });

  it('Should return a function for the onNavigation method', () => {
    expect(platformActions.onNavigation()).toMatchSnapshot('function');
  });

  it('Should return a dispatch object for the setAppName method', () => {
    expect(platformActions.setAppName()).toMatchSnapshot('dispatch object');
  });

  it('Should return a function for the setNavigation method', () => {
    expect(platformActions.setNavigation()).toMatchSnapshot('function');
  });
});
