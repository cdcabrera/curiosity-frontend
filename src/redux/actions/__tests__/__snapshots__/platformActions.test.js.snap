// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlatformActions Should return a dispatch object for the hideGlobalFilter method: dispatch object 1`] = `
{
  "payload": Promise {},
  "type": "PLATFORM_GLOBAL_FILTER_HIDE",
}
`;

exports[`PlatformActions Should return response content for createExport method: dispatch object 1`] = `
[
  [
    {
      "meta": {
        "id": undefined,
        "notifications": {
          "pending": {
            "dismissable": true,
            "id": "swatch-create-export-undefined",
            "title": "t(curiosity-toolbar.notifications_export_pending, {"context":"title"})",
            "variant": "info",
          },
          "rejected": {
            "description": "t(curiosity-toolbar.notifications_export_error, {"context":"description"})",
            "dismissable": true,
            "title": "t(curiosity-toolbar.notifications_export_error, {"context":"title"})",
            "variant": "warning",
          },
        },
      },
      "payload": Promise {},
      "type": "SET_PLATFORM_EXPORT_CREATE",
    },
  ],
]
`;

exports[`PlatformActions Should return response content for getExistingExports method: dispatch object 1`] = `
[
  [
    {
      "meta": {
        "notifications": {
          "fulfilled": {
            "description": "t(curiosity-toolbar.notifications_export_completed, {"context":"descriptionGlobal","count":0})",
            "dismissable": true,
            "title": "t(curiosity-toolbar.notifications_export_completed, {"context":"titleGlobal","count":0})",
            "variant": "success",
          },
          "pending": {
            "dismissable": true,
            "title": "t(curiosity-toolbar.notifications_export_pending, {"context":"titleGlobal"})",
            "variant": "info",
          },
          "rejected": {
            "description": "t(curiosity-toolbar.notifications_export_error, {"context":"description"})",
            "dismissable": true,
            "title": "t(curiosity-toolbar.notifications_export_error, {"context":"title"})",
            "variant": "warning",
          },
        },
      },
      "payload": Promise {},
      "type": "GET_PLATFORM_EXPORT_EXISTING",
    },
  ],
]
`;

exports[`PlatformActions Should return response content for getExistingExportsStatus method: dispatch object 1`] = `
[
  [
    {
      "meta": {
        "notifications": {
          "rejected": {
            "description": "t(curiosity-toolbar.notifications_export_error, {"context":"description"})",
            "dismissable": true,
            "title": "t(curiosity-toolbar.notifications_export_error, {"context":"title"})",
            "variant": "warning",
          },
        },
      },
      "payload": Promise {},
      "type": "SET_PLATFORM_EXPORT_EXISTING_STATUS",
    },
  ],
]
`;

exports[`PlatformActions Should return response content for removeExistingExports method: dispatch object 1`] = `
[
  [
    {
      "meta": {
        "notifications": {
          "rejected": {
            "description": "t(curiosity-toolbar.notifications_export_error, {"context":"description"})",
            "dismissable": true,
            "title": "t(curiosity-toolbar.notifications_export_error, {"context":"title"})",
            "variant": "warning",
          },
        },
      },
      "payload": Promise {},
      "type": "DELETE_PLATFORM_EXPORT_EXISTING",
    },
  ],
]
`;
